### PR TITLE
Add support for singleton runners

### PR DESCRIPTION
**Pull request by 0x2b3bfa0**
The --reuse-existing flag will cause cml-runner to look for already existing runners with the specified --labels and skip creating a new one if that's the case.

### CODE REVIEW

**Comment by DavidGOrtega** [comment]
Thinking about this PR... why labels over name? I think that name could be the best case here since its also represents one instead of several so the idea could be:
 - if name is in use, return 0 instead of throwing the exception like it is doing 
 - if name is in use and flag --dry or something alike would end up in an exception (maybe this is not even needed)

What do you think?

**Comment by 0x2b3bfa0** [comment]
@DavidGOrtega, I used labels because of this comment on Discord:

>[_Is there an elegant way to check if an appropriate self-hosted instance wit the appropriate tags is already available (if yes, skipping, if no, create one)?_](https://discord.com/channels/485586884165107732/728693131557732403/816001680004350004)

Nevertheless, the hacks I made at my command prompt for the sake of curiosity were using names instead of labels, and as you stated, the implementation would be a lot easier.

The main conceptual difference I see between labels and names is whether we want to deduplicate runners based on their functionality or their identity. I find the former more adequate from a purist standpoint, but can't find any practical case where the latter wouldn't be able to offer an identical behavior with the current implementation.

Are we planning to offer the possibility of launching a fixed number of identical runners with the same invocation? If so, it would make sense to choose labels over names, so we can count how many of them are listening for jobs. _Id est:_

```console
% cml-runner --count 3
Registered runner uniquename-1...
Registered runner uniquename-2...
Registered runner uniquename-3...
% cml-runner --count 3
Reusing runner uniquename-1...
Reusing runner uniquename-2...
Registered runner uniquename-3 again because the third runner died after being idle for ${timeout} seconds...
```

**Comment by DavidGOrtega** [comment]
> Are we planning to offer the possibility of launching a fixed number of identical runners with the same invocation?

We already support that. You can give a tf file to cml-runner command so you can actually launch a cluster through TF. 

> ... but can't find any practical case where the latter wouldn't be able to offer the same behavior with the current implementation

Thats what I feel too. Labels can act as a unique id or can as a group id.
Actually if we were logging instead of throwing using name instead of labels would work, even in. not adding in the workflow that the second job has to continue even after the failure.

**Comment by 0x2b3bfa0** [comment]
You added the following fragment after I [posted my reply](https://github.com/iterative/cml/pull/422#issuecomment-790941724), so I'll answer here to keep the conversation linearity:

> [_Thinking also about name... is name really needed? Indeed there is a bug in GL that we cant search by name as they state... I wonder if we could not simplify all by label... I mean name used internally to name the host and runner but the user is not able to set it (adding labels to the host also) and stick with labels_](https://github.com/iterative/cml/pull/422#issuecomment-790932716)

I would prefer this label-centric solution over the current name approach, given that we only care about the capabilities offered by runners and not their identity. Moreover, it would be consistent with the workflow syntax where users select runners by labels and not by name.

What if we deprecate `--name` with a warning and encourage the use of labels?

**Comment by DavidGOrtega** [comment]
> What if we deprecate --name with a warning and encourage the use of labels?
makes totally sense to me

**Comment by 0x2b3bfa0** [comment]
I've commited the warning before I forget, but can be rolled back if required.

File: bin/cml-runner.js [hunk 1]
```diff
@@ -287,6 +308,12 @@ const opts = decamelize(
     .boolean('single')
     .default('single', RUNNER_SINGLE)
     .describe('single', 'If specified, exit after running a single job.')
+    .boolean('reuse-existing')
+    .default('reuse-existing', RUNNER_REUSE_EXISTING)
+    .describe(
+      'reuse-existing',
```
Comment (DavidGOrtega) [inline bin/cml-runner.js:314]: ```reuse-existing``` I would make it shorter and also there is one implication with the possibility of launching several with a tf file that we have not even thought.

**Comment by DavidGOrtega** [comment]
Not sure In general... I think that we should deprecate ```name```  in the future. 
To do such thing I would change the thrown exception in the duplicated with a process.exit(0) plus log.

**Comment by 0x2b3bfa0** [comment]
> _Not sure In general... I think that we should deprecate name in the future._

Do you mean to postpone the warning message for a future version? That could be a perfectly valid option, and we could silently accept both functionalities for now.

> _To do such thing I would change the thrown exception in the duplicated with a process.exit(0) plus log._

Would that imply deduplicating runners by name instead of by labels, and just emitting a warning instead of an error like we're doing now?

**Comment by 0x2b3bfa0** [comment]
> _also there is one implication with the possibility of launching several with a tf file that we have not even thought_

Given that the `get_runners_by_labels` method returns a list of runners, we could count how many are there and use that as limit, but accounting for the flexibility of custom Terraform code could be an impossible goal.

**Comment by DavidGOrtega** [comment]
```js
if (await cml.runner_by_name({ name })) {
    throw new Error(
      `Runner name ${name} is already in use. Please change the name or terminate the other runner.`
    );
```

we should add here the reuse flag to not throw and log and exit instead

**Comment by DavidGOrtega** [comment]
@0x2b3bfa0 did you review my last comment?

**Comment by 0x2b3bfa0** [comment]
@DavidGOrtega, sorry for the delay, I was finishing a proposal for #217

File: bin/cml-runner.js [hunk 2]
```diff
@@ -256,11 +267,29 @@ const run = async (opts) => {
...
+  //   );
+  // }
+
   await cml.repo_token_check();
-  if (await cml.runner_by_name({ name }))
-    throw new Error(
-      `Runner name ${name} is already in use. Please change the name or terminate the other runner.`
-    );
+
+  if (reuse_existing && (await cml.runners_by_labels({ labels })).length > 0) {
```
Comment (0x2b3bfa0) [inline bin/cml-runner.js:278]: The `.length > 0` unnecessary comparison was just to exemplify how to keep a minimum number fo runners alive, but can be removed for the sake of cleanliness.

Comment (0x2b3bfa0) [reply bin/cml-runner.js:278]:
```suggestion
  if (reuse_existing && (await cml.runners_by_labels({ labels }))) {
```

File: bin/cml-runner.js [hunk 3]
```diff
@@ -256,11 +267,29 @@ const run = async (opts) => {
...
+    console.log(`Reusing existing runners with the ${labels} labels...`);
+    process.exit(0);
+  }
+
+  if (await cml.runner_by_name({ name })) {
+    if (reuse_existing) {
+      console.log(`Reusing existing runner named ${name}...`);
+      process.exit(0);
```
Comment (0x2b3bfa0) [inline bin/cml-runner.js:286]: @DavidGOrtega, is that what you wanted?

File: bin/cml-runner.js [hunk 4]
```diff
@@ -256,11 +267,29 @@ const run = async (opts) => {
...
-      `Runner name ${name} is already in use. Please change the name or terminate the other runner.`
-    );
+
+  if (reuse_existing && (await cml.runners_by_labels({ labels }))) {
+    console.log(`Reusing existing runners with the ${labels} labels...`);
+    process.exit(0);
+  }
+
+  if (await cml.runner_by_name({ name })) {
+    if (reuse_existing) {
```
Comment (DavidGOrtega) [inline bin/cml-runner.js:284]: Cleaner with: 

if (!reuse_existing) throw

log
exit

🙏

File: bin/cml-runner.js [hunk 5]
```diff
@@ -256,11 +267,29 @@ const run = async (opts) => {
...
+  //   );
+  // }
+
   await cml.repo_token_check();
-  if (await cml.runner_by_name({ name }))
-    throw new Error(
-      `Runner name ${name} is already in use. Please change the name or terminate the other runner.`
-    );
+
+  if (reuse_existing && (await cml.runners_by_labels({ labels }))) {
```
Comment (DavidGOrtega) [inline bin/cml-runner.js:278]: true && [] && console.log('here!')

Needs the length check

File: bin/cml-runner.js [hunk 6]
```diff
@@ -287,6 +316,12 @@ const opts = decamelize(
     .boolean('single')
     .default('single', RUNNER_SINGLE)
     .describe('single', 'If specified, exit after running a single job.')
+    .boolean('reuse-existing')
+    .default('reuse-existing', RUNNER_REUSE_EXISTING)
```
Comment (DavidGOrtega) [inline bin/cml-runner.js:320]: reuse or reuse-runner

File: bin/cml-runner.js (hunk 5 above)
Comment (0x2b3bfa0) [reply bin/cml-runner.js:278]: 🙏  Thanks! `[] && console.log("🤦 ")` Sorry for the lapsus, my brain was damaged by snake taming. 🐍

Comment (0x2b3bfa0) [reply bin/cml-runner.js]: Today I learned that `([] == false) === true` and `!![] == true` despite `Boolean([]) == true` 🤯

Comment (DavidGOrtega) [reply bin/cml-runner.js]: JS the good parts by Douglas Crockford. Highly recommended

**Comment by 0x2b3bfa0** [comment]
@DavidGOrtega, it seems to be working for everything but GitLab (untested).

File: bin/cml-runner.js [hunk 7]
```diff
@@ -256,11 +267,27 @@ const run = async (opts) => {
     await tf.save_tfstate({ tfstate, path });
   }
 
+  // if (name !== NAME) {
```
Comment (DavidGOrtega) [inline bin/cml-runner.js:270]: Please remove commented code

**Review by DavidGOrtega — APPROVED** [review]

File: bin/cml-runner.js [hunk 8]
```diff
@@ -256,11 +267,27 @@ const run = async (opts) => {
...
+  //   console.log(
+  //     'The --name option is deprecated: please use --labels to differentiate runners instead.'
+  //   );
+  // }
+
```
Comment (0x2b3bfa0) [inline bin/cml-runner.js:275]:
```suggestion
```

**Pull request merged**